### PR TITLE
Add GroupBy() Filter

### LINKED ISSUE #1726: add filter to Group By
### Description

It should be possible to apply an arbitrary filter to Group By queries. I think the fact that it isn't there is basically an oversight on my part. If I'm thinking about this correctly, it should be possible to Intersect the filter with each value of the first field in the group by, and basically add almost no overhead.

Difficult question will be designing the API. We could make the filter a keyword arg.
`GroupBy(Rows(field=blah), Rows(field=blah2), limit=10, filter=Row(field=bar, id=2))`

Or an optional first positional arg:
`GroupBy(Row(field=bar, id=2), Rows(field=blah), Rows(field=blah2), limit=10)`

which would require a little extra care in processing to check if the first query is a bitmap type query or a rows type query, and might be a bit more bug prone. Also somewhat more confusing, although follows how TopN works...

### Success criteria (What criteria will consider this ticket closeable?)
Group By with filter (basically a `WHERE... GROUP BY ...` ) is implemented.

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds the ability to specify a row call as the first argument to `GroupBy()` and apply that as a filter to the first level `GroupBy()` results.

Fixes #1726

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

File: executor.go [hunk 1]
```diff
@@ -874,8 +874,15 @@ func (r RowIDs) merge(other RowIDs, limit int) RowIDs {
 }
 
 func (e *executor) executeGroupBy(ctx context.Context, index string, c *pql.Call, shards []uint64, opt *execOptions) ([]GroupCount, error) {
+	// extract filter call.
+	var filter *pql.Call
+	children := c.Children
+	if len(children) > 0 && isBitmapCallName(children[0].Name) {
+		filter, children = children[0], children[1:]
+	}
```
Comment (benbjohnson) [inline executor.go]: This splits off the `filter` from the rest of the `Rows()` children. These are propagated through the shard calls later.

File: executor.go [hunk 2]
```diff
@@ -1028,8 +1035,15 @@ func (g GroupCount) Compare(o GroupCount) int {
...
 
-func (e *executor) executeGroupByShard(_ context.Context, index string, c *pql.Call, shard uint64, childRows []RowIDs) ([]GroupCount, error) {
-	iter, err := newGroupByIterator(childRows, c.Children, index, shard, e.Holder)
+func (e *executor) executeGroupByShard(ctx context.Context, index string, c *pql.Call, children []*pql.Call, filter *pql.Call, shard uint64, childRows []RowIDs) (_ []GroupCount, err error) {
+	var filterRow *Row
+	if filter != nil {
+		if filterRow, err = e.executeBitmapCallShard(ctx, index, filter, shard); err != nil {
+			return nil, errors.Wrapf(err, "executing group by filter for shard %d", shard)
+		}
+	}
```
Comment (benbjohnson) [inline executor.go]: Filter is executed per-shard and passed to the `groupByIterator` below.

File: executor.go [hunk 3]
```diff
@@ -2756,6 +2774,11 @@ func newGroupByIterator(rowIDs []RowIDs, children []*pql.Call, index string, sha
 		}
 	}
 
+	// Apply filter to first level, if available.
+	if gbi.filter != nil && len(gbi.rows) > 0 {
+		gbi.rows[0].row = gbi.rows[0].row.Intersect(gbi.filter)
+	}
```
Comment (benbjohnson) [inline executor.go]: Applies the filter to the first level on the initializer.

File: executor.go [hunk 4]
```diff
@@ -2774,10 +2797,12 @@ func (gbi *groupByIterator) nextAtIdx(i int) {
...
-	if i != 0 && i != len(gbi.rows)-1 {
-		gbi.rows[i].row = nr.Intersect(gbi.rows[i-1].row)
-	} else {
+	if i == 0 && gbi.filter != nil {
+		gbi.rows[i].row = nr.Intersect(gbi.filter)
+	} else if i == 0 || i == len(gbi.rows)-1 {
 		gbi.rows[i].row = nr
+	} else {
+		gbi.rows[i].row = nr.Intersect(gbi.rows[i-1].row)
 	}
```
Comment (benbjohnson) [inline executor.go]: Applies the filter to the first level on each subsequent call to the iterator.

**Comment by benbjohnson** [comment]
I think the `filter` keyword is probably better. It seemed like it could get hairy so I pushed in the basic filter functionality first. I'll add the PQL parser change and refactor.

File: pql/ast.go [hunk 5]
```diff
@@ -329,6 +343,22 @@ func (c *Call) UintSliceArg(key string) ([]uint64, bool, error) {
...
+func (c *Call) CallArg(key string) (*Call, bool, error) {
+	val, ok := c.Args[key]
+	if !ok {
+		return nil, false, nil
+	}
+	switch tval := val.(type) {
+	case *Call:
+		return tval, true, nil
+	default:
+		return nil, true, fmt.Errorf("could not convert %v of type %T to Call in Call.UintArg", tval, tval)
```
Comment (jaffee) [inline pql/ast.go]: "...in Call.CallArg" (copy/paste issue)

**Comment by benbjohnson** [comment]
@jaffee Fixed up the nits. Ready for re-review.

**Pull request merged**